lang/py/src/avro/io.py

```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
-    return isinstance(datum, str) and len(datum) == expected_schema.size
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
+        BASE_ERR.format('logicalType(Decimal), baseType(fixed)', type(datum))
+      )
+    if not isinstance(datum, str):
+      return ValidationResult(BASE_ERR.format('str', type(datum)))
+    elif len(datum) != expected_schema.size:
+      return ValidationResult(
+        'Fixed type with invalid size, expected "{}" bytes found "{}" bytes'.format(
+          expected_schema.size, len(datum)
+        )
+      )
+    else:
+      return ValidationResult()
   elif schema_type == 'enum':
-    return datum in expected_schema.symbols
+    return ValidationResult(
+      None if datum in expected_schema.symbols else \
+      '"{}" is not among expected symbols {}'.format(datum, expected_schema.symbols)
+    )
   elif schema_type == 'array':
-    return (isinstance(datum, list) and
-      False not in [validate(expected_schema.items, d) for d in datum])
+    if not isinstance(datum, list):
+      return ValidationResult(BASE_ERR.format('list', type(datum)))
+    arr_items = [validate(expected_schema.items, d) for d in datum]
+    if not any(arr_items):
+      return ValidationResult()
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, arr_items)]))
   elif schema_type == 'map':
-    return (isinstance(datum, dict) and
-      False not in [isinstance(k, basestring) for k in datum.keys()] and
-      False not in
-        [validate(expected_schema.values, v) for v in datum.values()])
+    if not isinstance(datum, dict):
+      return ValidationResult(BASE_ERR.format('dict', type(datum)))
+    if False in [isinstance(k, basestring) for k in datum.keys()]:
+      return ValidationResult('Map keys must be string typed')
+    val_items = [validate(expected_schema.values, v) for v in datum.values()]
+    if not any(val_items):
+      return ValidationResult()
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, val_items)]))
   elif schema_type in ['union', 'error_union']:
-    return True in [validate(s, datum) for s in expected_schema.schemas]
+    if not all([validate(s, datum) for s in expected_schema.schemas]):
+      return ValidationResult()
+    return ValidationResult('"{}" is not an example of any union schema'.format(datum))
   elif schema_type in ['record', 'error', 'request']:
-    return (isinstance(datum, dict) and
-      False not in
-        [validate(f.type, datum.get(f.name)) for f in expected_schema.fields])
+    if not isinstance(datum, dict):
+      return ValidationResult(BASE_ERR.format('dict', type(datum)))
+    err_info = []
+    for f in expected_schema.fields:
+      res = validate(f.type, datum.get(f.name))
+      if not res:
+        if datum.get(f.name):
+          err_info.append('"{}": {}'.format(f.name, res.msg))
+        else:
+          err_info.append('"{}": missing and not nullable.'.format(f.name))
```

Comment:
The version I just pushed will simply output:
`"message": missing`
if the key is item is invalid and also not present in the provided data. Luckily if the key is nullable, it won't be invalid in the first place since it'll validate against the union ['string', 'null'] or whatnot.